Plumb symlink support through the Pants engine

thejcannon opened the pull request:
This plumbs everything needed to support symlinks through the engine without actually having them be picked up by path globbing (which will come later)

thejcannon commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -499,29 +548,43 @@ impl DigestTrie {
     digests
   }
 
-  /// Return a pair of Vecs of the file paths and directory paths in this DigestTrie, each in
-  /// sorted order.
-  ///
-  /// TODO: This should probably be implemented directly by consumers via `walk`, since they
-  /// can directly allocate the collections that they need.
-  pub fn files_and_directories(&self) -> (Vec<PathBuf>, Vec<PathBuf>) {
```
I split this into 3 implementations, as only one caller needed all the items at the same time, and it was for `__repr__`, so perf doesn't matter __that__ much.

thejcannon reviewed (commented):
@tdyas or @stuhood can y'all comment on the `todo!()`s? Anything from actual code to copy+paste to high-level description of how to handle that case would be very nice :smile:

tdyas commented on src/python/pants/engine/fs.py:
```diff
@@ -73,6 +73,14 @@ class FileEntry:
 
 
+@dataclass(frozen=True)
+class LinkEntry:
```
This should probably be called `SymlinkEntry` for consistency with usages of "symlink" elsewhere (including in the REAPI where there is a `SymlinkNode` to represent them).

stuhood replied:
Meanwhile, we use `Link` throughout `fs.rs`.

tdyas commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -311,6 +314,43 @@ impl From<&File> for remexec::FileNode {
+    self.name
+  }
+
+  pub fn target(&self) -> String {
```
Generally this should return `&str` so that callers are not forced to take ownership (and incur an allocation) if they merely want to examine the target. The caller can always call `.to_owned()` itself if it needs ownership.

tdyas replied:
See comment about typing of `target` as a `RelativePath` or `PathBuf`.

tdyas commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -311,6 +314,43 @@ impl From<&File> for remexec::FileNode {
+#[derive(Clone, Debug, DeepSizeOf)]
+pub struct Symlink {
+  name: Name,
+  target: String,
```
Maybe `target` should be typed as `RelativePath` or `PathBuf`? `RelativePath` is a Pants concept that enforces that the path is a relative path.

For now, I don't believe we need to support absolute paths (other than maybe `/dev/null`).

thejcannon replied:
Went with `RelativePath`

stuhood commented on src/rust/engine/fs/store/src/lib.rs:
```diff
@@ -39,8 +39,8 @@ use std::collections::{BTreeMap, HashMap, HashSet};
 use std::future::Future;
-use std::io::{self, Read, Write};
-use std::os::unix::fs::{OpenOptionsExt, PermissionsExt};
+use std::io::{self, Error, Read, Write};
```
Rather than importing `Error`, this `self` import allows you to refer to it as `io::Error`, which is preferred when something is potentially ambiguous like this.

stuhood commented on src/rust/engine/fs/store/src/lib.rs:
```diff
@@ -941,6 +949,7 @@ impl Store {
       .await?
       .walk(&mut |_, entry| match entry {
         directory::Entry::File(f) => file_digests.push(f.digest()),
+        directory::Entry::Symlink(_) => todo!(),
         directory::Entry::Directory(_) => (),
```
Nothing to do here... if the symlink is relative within the tree, then it will already be visited: if it is outside the tree then there is nothing to do.

```suggestion
        directory::Entry::Symlink(_) | directory::Entry::Directory(_) => (),
```

stuhood commented on src/rust/engine/fs/store/src/lib.rs:
```diff
@@ -901,6 +908,7 @@ impl Store {
           file_digests.insert(f.digest());
         }
         directory::Entry::Directory(_) => (),
+        directory::Entry::Symlink(_) => todo!(),
```
Ditto.

```suggestion
        directory::Entry::Directory(_) | directory::Entry::Symlink(_) => (),
```

tdyas replied:
Do nothing here since the goal of this loop is to find file digests so that those digests can be checked for existence. Symlinks always "exist" without such further checks since the symlink exists if and only if it exists as a `SymlinkNode` in a `Directory` proto.

The goal of this method is ensure that file and directory digests are present in the CAS.

tdyas commented on src/rust/engine/fs/src/lib.rs:
```diff
@@ -766,16 +790,24 @@ pub struct FileEntry {
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub struct LinkEntry {
```
`SymlinkEntry`

tdyas commented on src/rust/engine/fs/store/src/lib.rs:
```diff
@@ -1301,6 +1311,11 @@ impl Store {
       .await?
   }
 
+  async fn materialize_link(&self, destination: PathBuf, target: String) -> Result<(), StoreError> {
```
`materialize_symlink`

stuhood commented on src/rust/engine/fs/store/src/lib.rs:
```diff
@@ -1314,6 +1329,7 @@ impl Store {
       .await?
       .walk(&mut |path, entry| match entry {
         directory::Entry::File(f) => files.push((path.to_owned(), f.digest(), f.is_executable())),
+        directory::Entry::Symlink(_) => (),
```
Sooo, this one is tricky. This is asking for `Files`, which would be symlink oblivious. So I think that you'd need to actually walk the symlink in this case to determine whether the destination was a file.

But also: if the destination is a `Directory`, then you'd need to recurse with a new relative path.

----

I see two options for these cases (where a `DigestTrie` needs to actually "act" like a filesystem):
1. adding another implementation of `DigestTrie::walk` that is symlink oblivious (i.e. will never return a symlink, and instead will walk within the destination of the symlink)
2. Adding methods to `VFS` to implement walking, and/or using the equivalent of `vfs.expand(PathGlobs(["**"]))`, and using them when you want to be symlink oblivious.

The former might be easier? It would be lightly redundant with path globs expansion, but on the other hand, pretending that something is POSIX when it isn't doesn't do you many favors.

thejcannon replied:
I added a param to `walk`, which should force callers to think about which behavior they want.

thejcannon replied:
Hmmm there's also the issue of "if its symlink oblivious and the symlink points outside of the trie"

tdyas replied:
Seems fine to me to allow dangling symlinks. The target of the symlink is the responsibility of the code invoking the `Process`. The VFS code should just do what it is told to do since it has no semantic knowledge of what the `Digest` is being used for.

stuhood replied:
On the other hand, it could be good to be defensive about this until someone identifies a use-case where they need a dangling symlink. The spec supports them, but leaving them as a TODO that we error for rather than partially implemented seems safer.

thejcannon replied:
Welllllll... I actually have a use-case :joy: 

I'm unpacking libs where one has a symlink to a lib in a dependency. So in reality the digests for libA will be merged with its dependencies and the link won't dangle. But in isolation its a dangling link.

In this case, we get a choice If the user tell us they are `oblivious` we could skip dangling symlinks or still return the symlink (so "oblivious" means "follow if possible"). I err on the side of "skipping"

stuhood replied:
Sorry: "dangling but pointing inside of the snapshot root" is one thing, and probably fine.

But "pointing outside of the root of the snapshot", either by being absolute, or doing a relative path lookup outside of the root is what I'm referring to (trying to) ban initially.

thejcannon replied:
In this case it'd be relative to the root, but the target is not in the Trie.

thejcannon replied:
In general though this code only needs to handle symlinks whose targets are not entries in the Trie.

I'm pretty sure other code handles parsing the target as relative and relative to sandbox root.

tdyas replied:
>But "pointing outside of the root of the snapshot", either by being absolute, or doing a relative path lookup outside of the root is what I'm referring to (trying to) ban initially.

Agreed. We should ban symlink targets escaping the sandbox.

stuhood replied:
> > But "pointing outside of the root of the snapshot", either by being absolute, or doing a relative path lookup outside of the root is what I'm referring to (trying to) ban initially.
> 
> Agreed. We should ban symlink targets escaping the sandbox.

Yep.

> I'm pretty sure other code handles parsing the target as relative and relative to sandbox root.

Yea... I think that as long as the snapshot capture code continues to not support a symlink escaping the root (`VFS::expand` -> `Snapshot::from_path_stats`), we're good.

thejcannon replied:
Marking as resolved. Added the param to `walk` for aware/oblivious and made each caller aware (for now, let's visit them inline)

tdyas commented on src/rust/engine/fs/store/src/lib.rs:
```diff
@@ -941,6 +949,7 @@ impl Store {
       .await?
       .walk(&mut |_, entry| match entry {
         directory::Entry::File(f) => file_digests.push(f.digest()),
+        directory::Entry::Symlink(_) => todo!(),
```
Do nothing here. Symlinks are always present since there is no separate digest to load for them.

stuhood commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -311,6 +314,43 @@ impl From<&File> for remexec::FileNode {
+    self.name
+  }
+
+  pub fn target(&self) -> String {
+    self.target.to_owned()
```
Generally you should expose a reference and let the caller decide whether they want to clone/to_owned it.
```suggestion
  pub fn target(&self) -> &str {
    &self.target
```

tdyas commented on src/rust/engine/fs/store/src/snapshot_ops.rs:
```diff
@@ -227,6 +238,7 @@ pub trait SnapshotOps: Clone + Send + Sync + 'static {
       directory::Entry::File(f) => {
         files.insert(path.to_owned(), f.digest());
       }
+      directory::Entry::Symlink(_) => todo!(),
```
Any change here depends on what (if any) changes are needed to the signature of `DigestTrie::from_path_stats` to support symlinks.

tdyas commented on src/rust/engine/process_execution/src/remote_cache.rs:
```diff
@@ -129,8 +129,9 @@ impl CommandRunner {
-      Some(directory::Entry::Directory(d)) => d.tree(),
       None => return Ok(None),
+      Some(directory::Entry::Directory(d)) => d.tree(),
+      Some(directory::Entry::Symlink(_)) => todo!(),
```
Error here for same reasons as this errors for files. An output _directory_ was expected but it was not a directory.

tdyas commented on src/rust/engine/process_execution/src/remote_cache.rs:
```diff
@@ -144,6 +145,7 @@ impl CommandRunner {
     let mut file_digests = Vec::new();
     sub_trie.walk(&mut |_, entry| match entry {
       directory::Entry::File(f) => file_digests.push(f.digest()),
+      directory::Entry::Symlink(_) => todo!(),
```
Do nothing, this loop gathers file digests that the caller will need to do something with. Symlinks don't have a digest.

tdyas commented on src/rust/engine/process_execution/src/remote_cache.rs:
```diff
@@ -164,7 +167,7 @@ impl CommandRunner {
         Ok(Some(output_file))
       }
-      None => Ok(None),
+      Some(directory::Entry::Symlink(_)) => todo!(),
```
Error here for same reason as directory variant errors. An output _file_ was expected and no file was found.

tdyas commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -893,6 +1019,7 @@ impl From<&DigestTrie> for remexec::Tree {
     trie.walk(&mut |_, entry| {
       match entry {
         Entry::File(_) => (),
+        Entry::Symlink(_) => todo!(),
```
Do nothing. This loop gathers Directory protos so they can be consolidated into a `Tree` proto. (Metadata for files and symlinks will be stored in those `Directory` protos.)

stuhood commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -311,6 +314,43 @@ impl From<&File> for remexec::FileNode {
+
 // TODO: `PathStat` owns its path, which means it can't be used via recursive slicing. See
 // whether these types can be merged.
 enum TypedPath<'a> {
```
The fact that this code hasn't changed suggests that `PathStat`, which is the type produced while `VFS::expand` walks the filesystem has not been changed.

If you get tests passing without that change, then I _think_ that what you will have accomplished is that our VFS code (which is what matches globs in the repo for `PathGlobs`->`Digest`, and which [captures the outputs of processes from sandboxes](https://github.com/pantsbuild/pants/blob/f72b8c0366aba7fa0a16bb83d49983113e8e44af/src/rust/engine/process_execution/src/local.rs#L491-L518)) will never actually create a `DigestTrie`/`Digest`/`Snapshot` containing a symlink, BUT when one was created synthetically, it would present as such in a memory, and be successfully materialized.

That's probably a good breaking point for a first PR, so you might want to avoid dipping into `VFS` yet if you can.

thejcannon replied:
Yeah just started to muck with `fs_test` and noticed this behavior.

I like the idea of splitting into multiple PRs.

stuhood reviewed (commented):
Nice!

tdyas commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -621,7 +621,20 @@ impl DigestTrie {
       let path = path_so_far.join(entry.name().as_ref());
-      f(&path, entry);
+
+      match (symlink_behavior.clone(), entry) {
```
No need for the `.clone`, just have the enum derive `Copy`. Or just match on a reference here, i.e., `match (&symlink_behavior, entry)`, and you still use the existing match arm as-is because Rust will infer a dereference  as necessary to make the match arm work. See docs on "match binding modes", https://doc.rust-lang.org/stable/book/ch18-01-all-the-places-for-patterns.html.

tdyas commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -621,7 +621,20 @@ impl DigestTrie {
+      };
+
       if let Entry::Directory(d) = entry {
         d.tree.walk_helper(path, symlink_behavior.clone(), f);
```
This clone is also unnecessary if the enum derives `Copy` (which it can because the enum variants don't carry data).

thejcannon commented:
Opening for review. LMK if we should tackle the `path_stats` in this PR, or wait for follow-up.

tdyas reviewed (commented):
> Opening for review. LMK if we should tackle the path_stats in this PR, or wait for follow-up.

I'm fine with the `from_path_stats` changes going to a separate PR. I would like to see some tests in this PR for the changes made in this PR.

stuhood commented on src/rust/engine/src/externs/fs.rs:
```diff
@@ -229,6 +242,18 @@ impl PySnapshot {
   }
 
+  #[getter]
+  fn symlinks<'py>(&self, py: Python<'py>) -> &'py PyTuple {
```
So, thinking about this more (in the context of being [symlink oblivious](https://github.com/pantsbuild/pants/issues/10271#issuecomment-1247086217)).

Currently `Snapshot` is symlink oblivious, in that users don't have to care about whether something is a symlink... they can view whatever was behind the symlink as a `Directory` or `File`. It's possible that even if `DigestTrie`/`Digest` have become symlink aware, that `Snapshot` should be symlink oblivious, for ease of use.

That's definitely subtle though, because it would probably only be `Snapshot` and `DigestContents` which would want to be symlink oblivious. `CreateDigest` and `DigestEntries` need to maintain their symmetry, and they should definitely gain the ability to manipulate symlinks in this PR.

thejcannon replied:
Just to make sure I understand. The `Snapshot.digest` would still contain the symlink information, but the properties `files` and `dirs` are symlink oblivious?

tdyas replied:
It doesn't need to be done in this PR but we could use a similar approach to what is done for `Process` -> `FallibleProcessResult` -> `ProcessResult`. That is, introduce a new `RawSnapshot` type (name TBD) that contains symlinks (along with files and directories) and converts to `Snapshot` by being symlink oblivious via new core rule.

stuhood replied:
> Just to make sure I understand. The `Snapshot.digest` would still contain the symlink information, but the properties `files` and `dirs` are symlink oblivious?

That's right.

> It doesn't need to be done in this PR but we could use a similar approach to what is done for `Process` -> `FallibleProcessResult` -> `ProcessResult`. That is, introduce a new `RawSnapshot` type (name TBD) that contains symlinks (along with files and directories) and converts to `Snapshot` by being symlink oblivious via a new core rule that does the conversion.

We could, but I think that `Snapshot` vs `DigestEntries` might already represent sufficiently high/low level APIs to make having two kinds of snapshot unnecessary?

tdyas replied:
> We could, but I think that Snapshot vs DigestEntries might already represent sufficiently high/low level APIs to make having two kinds of snapshot unnecessary?

Maybe? Do we see any users of `Snapshot` wanting to be symlink aware?

stuhood replied:
> Maybe? Do we see any users of `Snapshot` that would want to be symlink aware?

I can't really imagine what the use for it would be. Just knowing that a symlink exists (as a string name via `fn links() -> Vec<String>`) isn't really enough to do anything with it... you need to know the type.

tdyas replied:
`links` could return a list of (SOURCE, TARGET) tuples. So I agree with your point about this API being sufficient for now. If we don't see an immediate need for `Snapshot` users to need such information, then easy enough to defer on providing a solution given `DigestEntries` can be used to get the symlinks for any callers that need it.

stuhood replied:
> So I agree with your point about this API being sufficient for now.

To be clear: this API is not quite right at the moment: see https://github.com/pantsbuild/pants/pull/16844#discussion_r971274624

The `files`/`directories` methods of `PySnapshot` would need to be `SymlinkBehavior::Oblivious`, in order to return the files/dirs behind symlinks.

stuhood commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -499,34 +556,53 @@ impl DigestTrie {
+    files
+  }
+
+  /// Returns the files in the trie, not including directories pointing to files.
```
Mismatched.

thejcannon commented:
FYI picking this back up. Give me a bit to refresh myself :sweat_smile:

thejcannon commented:
It works! And I have a test for it too!

stuhood commented on src/python/pants/engine/fs.py:
```diff
@@ -73,6 +73,14 @@ class FileEntry:
 
+@dataclass(frozen=True)
+class SymlinkEntry:
+    """A symlink pointing to a target path."""
```
A bit more explanation of the constraints of this type (possibly by referring to the REAPI spec) might be good. Does the target need to be absolute, normalized, etc?

stuhood commented on src/python/pants/engine/fs.py:
```diff
@@ -96,15 +104,15 @@ class DigestContents(Collection[FileContent]):
+class DigestEntries(Collection[Union[FileEntry, SymlinkEntry, Directory]]):
     """The indirect file contents of a Digest.
 
     DigestEntries is a collection of FileContent and Directory instances representing, respecively,
```
Both of these types need doc updates. It might also be worth updating the docstring for `Snapshot` to point out the contrast with these types (symlink oblivious vs aware).

stuhood commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -539,15 +622,29 @@ impl DigestTrie {
+
+      match (&symlink_behavior, entry) {
+        (SymlinkBehavior::Oblivious, Entry::Symlink(s)) => {
+          if let Ok(Some(entry)) = self.entry(&s.target) {
+            f(&path, entry);
+          };
```
So, I _think_ that this needs to trigger recursion of `walk_helper`. Because if the destination is a directory, the effect of walking a symlink should actually be to continue walking within the destination, but with a new `path_so_far`.

Currently, it's only using the `entry` to call the user function, but it also needs to use that entry to continue recursing in `walk_helper` below.

To test that, you'd want to validate that:
```python
CreateDigest([
  FileEntry("a/1.txt", ..)
  SymlinkEntry("b", "a"),
])
```
... expands to the files:
```python
[
  "a/1.txt",
  "b/1.txt",
]
```

stuhood commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -952,6 +1123,9 @@ fn paths_of_child_dir(name: Name, paths: Vec<TypedPath>) -> Vec<TypedPath> {
           is_executable,
         },
+        TypedPath::Link { .. } => {
+          todo!();
```
TODO

stuhood commented on src/rust/engine/fs/store/src/lib.rs:
```diff
@@ -1327,13 +1349,14 @@ impl Store {
-      .await?
-      .walk(&mut |path, entry| match entry {
+    self.load_digest_trie(digest).await?.walk(
+      SymlinkBehavior::Aware,
```
Is `DigestContents` supposed to be symlink aware or symlink oblivious? I think that a consumer would want to pretend that symlinks do not exist, and get the content of files multiple times in case of symlinks, at least in part because that is symmetrical with `Snapshot`. So this should probably be oblivious.

stuhood reviewed (commented):
Thanks a lot! Two issues I think, then this looks ready to land =)

thejcannon commented:
OK I fixed the comments, and added new tests for additional behavior.

I'm struggling to debug why one aspect fails. These symlinks don't seem to be in the `.files`

```
                    SymlinkEntry("a/followed.ln", "file.txt"),
                    SymlinkEntry("a/followed-dotslash.ln", "./file.txt"),
```

thejcannon commented:
I also realized we don't handle `../` at all :face_exhaling:

tdyas commented:
> I also realized we don't handle `../` at all 😮‍💨

In what way?

thejcannon commented:
The symlink oblivious DigestTrie traversal doesn't do anything special for parent dirs, so it'll always count as as a "miss"

tdyas commented:
> The symlink oblivious DigestTrie traversal doesn't do anything special for parent dirs, so it'll always count as as a "miss"

To clarify, for the following `Digest`, it would not handle the `bar/hello.txt` symlink?

```
foo/hello.txt
bar/hello.txt -> ../foo/hello.txt
```

thejcannon commented:
Not yet no. I'm guessing we'll have to handle it by also keeping the Trie root around and just resolving symlinks from the root using the fully resolved canonical path.

Although at this point im sure there's corner cases on corner cases 😔

tdyas commented:
> Not yet no. I'm guessing we'll have to handle it by also keeping the Trie root around and just resolving symlinks from the root using the fully resolved canonical path.
> 
> Although at this point im sure there's corner cases on corner cases 😔

I'm fine if we solve that (and other edge cases) in subsequent PRs. Maybe add an expected-to-fail test for the use case and open an issue?

thejcannon commented:
I can dig it

stuhood commented:
>> I'm fine if we solve that (and other edge cases) in subsequent PRs. Maybe add an expected-to-fail test for the use case and open an issue?
>
> I can dig it

Hm... I'm less of a fan... I'd rather not land with known bugs (which we _could_ actually newly hit with this PR if a remote execution implementation returns symlinks).

thejcannon commented:
I can dig that too :sweat_smile:

thejcannon commented:
OK we're up to snuff on our subdirectory symlinks thanks to using `PathBuf` instead of `RelativePath`. Now we need to handle "up" dirs :smile:

tdyas commented on src/python/pants/engine/fs.py:
```diff
@@ -96,16 +114,17 @@ class DigestContents(Collection[FileContent]):
-    DigestEntries is a collection of FileContent and Directory instances representing, respecively,
-    actual files and empty directories present in the Digest.
+    DigestEntries is a collection of FileEntry/SymlinkEntry/Directory instances representing,
+    respecively, actual files, actual symlinks, and empty directories present in the Digest.
```
respecively -> respectively

thejcannon commented:
Latest changes support "updirs" (oh holy holy holy moly that hurt).

Please allow me to explain the madness as best I can :smile: 

First, when we _don't_ hit an "updir" in a symlink, we just traverse like normal through the `DigestTrie` (E.g. loosely, look for the symlink target in `self`, and keep continuing the work using the result). (This is in `tree_for_entry`).

So, now we've hit an "updir". There's 2 places this matters:
- `entry_helper`: This means we hit a path with an updir somewhere in the middle. This is handled by just normalizing the path during traversal. At this point we know that's "safe" because we've made our way down to this dir, we can just make our way back up.
- `tree_for_entry`: Here's the real magic. We only see an updir in this codeblock if we're looking to find the `tree` for a symlink target. For every `../` we go up an ancestor (but don't normalize, we don't know what other symlinks might be relevant once we arrive at our destination). Then we can join our ancestor and the remaining part of the symlink target and start traversing from the root all over again.

The result is the code for handling these things is relatively isolated, and only requires the `root` being passed around (instead of vectors of ancestor trees).

thejcannon commented:
Absolute paths still not handled :smile:

thejcannon commented on src/python/pants/engine/fs_test.py:
```diff
@@ -443,6 +444,118 @@ def test_digest_entries_handles_empty_directory(rule_runner: RuleRunner) -> None
+            id="symlink-makes-updir-valid",
+        ),
+    ],
+)
```
I'll be happy with A LOT more test cases

tdyas replied:
Do any of the other REAPI-implementing build tools have test cases to "borrow"?

thejcannon replied:
Ooohhh good idea

thejcannon replied:
Although, [what an abomination of code.](https://github.com/buildbarn/bb-remote-execution/blob/71af939fd205d5411aa6d4ae89d786dc9363cd84/pkg/builder/output_hierarchy_test.go#L257) It's more abstract art than code.

stuhood commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -539,19 +617,72 @@ impl DigestTrie {
+
+        if let Ok(Some(target_entry)) = tree.entry_helper(root, &target, link_depth) {
+          if link_depth >= MAX_LINK_DEPTH {
+            // We don't return a Result, so log and move on
+            warn!("Exceeded the maximum link depth while traversing links. Stopping traversal.");
```
This should probably include some information on the path that was traversed?

thejcannon replied:
I'm guessing we'd want to trim the path. Suggestion on some Rust-y code that'd do that?

stuhood reviewed (approved):
Thanks a lot!

Regarding absolute symlinks: it should just amount to ignoring symlinks dests which are: `Path::new(...).is_absolute()` (https://doc.rust-lang.org/nightly/std/path/struct.Path.html#method.is_absolute) in `walk`?

thejcannon commented:
Yeah I just hit the buzzer before I implemented absolute paths. It's up next.

thejcannon commented:
Absolute paths are donezo. Only problem I think is a symlink to a single `"."`. I thin it needs to be handled very specially, possibly in multiple places (where we know what the preceeding path comp is.

thejcannon commented:
Ah I think it's a combo of handling a symlink with a bare ".", And peeking for a "." In the components loop in entry_helper.

thejcannon commented:
Well that was painful...

thejcannon commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -539,16 +617,52 @@ impl DigestTrie {
+      let mut entry = entry;
+      if let SymlinkBehavior::Oblivious = symlink_behavior {
+        if let Entry::Symlink(s) = entry {
+          if link_depth >= MAX_LINK_DEPTH {
+            warn!("Exceeded the maximum link depth while traversing links. Stopping traversal.");
+            return;
+          }
```
This technically only exists for the `CurDir` case, as we'd never use the `entry_helper`, and therefore never hit it's link depth check :/

tdyas replied:
Worth a comment with that point?

thejcannon replied:
Yeah, now that I think about just moving the check into the curdir block would make that obvious

thejcannon commented on src/rust/engine/fs/src/directory.rs:
```diff
@@ -745,42 +879,88 @@ impl DigestTrie {
       path_so_far.push(component);
-      let component = component.as_os_str();
+      logical_path.push(component);
+      if component == Component::ParentDir {
```
This very well could be a `while` loop, but the perf seems negligible and I'm tired of staring at the code.

stuhood commented on src/rust/engine/fs/src/directory_tests.rs:
```diff
@@ -0,0 +1,372 @@
+use crate::MAX_LINK_DEPTH;
+use hashing::EMPTY_DIGEST;
+use std::path::{Path, PathBuf};
+
```
Thanks for doing this... I'm lightly embarassed not to have had rust side tests for `walk`/`entry` already... but on the other hand, it used to be a _lot_ simpler, heh.

thejcannon replied:
Having these tests is the only way I survived honestly. Turnaround time for changes went from minutes to seconds, and the ability to log/debug easily was invaluable.

stuhood reviewed (approved):
Thanks a lot for persevering here, and for being so thorough with test coverage!

thejcannon commented:
> Thanks a lot for persevering here, and for being so thorough with test coverage!

I was __not__ about to half-ass this. Especially because bugs here would absolutely wreck havoc on plugin authors and the end-user.

thejcannon commented:
Hmm wasn't expecting `test_digest_entries_handles_symlinks` to timeout :exclamation:

thejcannon commented:
Oh one thing to note is when we're walking a Trie as symlink oblivious, directory symlinks are not passed through the function as directories.

E.g. self/foo.bar might be an entry for a file file while self is not a directory entry.

Cc @stuhood

thejcannon commented on src/rust/engine/fs/src/directory_tests.rs:
```diff
@@ -0,0 +1,372 @@
+}
+
+#[test]
+fn walk_too_many_links_subdir() {
+  let tree = make_tree(vec![
+    TypedPath::File {
+      path: Path::new("a/file.txt"),
+      is_executable: false,
+    },
+    TypedPath::Link {
+      path: Path::new("a/self"),
+      target: Path::new("."),
+    },
+  ]);
+  assert_walk(
+    &tree,
+    (0..MAX_LINK_DEPTH)
+      .into_iter()
+      .map(|n| ("a/".to_string() + &"self/".repeat(n.into()) + "file.txt"))
+      .collect::<Vec<_>>(),
+    vec!["".to_string(), "a".to_string()],
```
Oh yeah @stuhood here's what I'm talking about. `a/self/file.txt` is a file (entry is for `a/file.txt`, but the only directory entries are root and `a`. 

I think conceptually it makes sense. There is no entry for `a/self`. But an interesting quirk.
